### PR TITLE
Libgit2 usability issues

### PR DESCRIPTION
**Pull request by Keno**
This is WIP, because the test fails non-deterministically and I've been unable to debug (was trying to use rr to figure out what's happening, but it crashes rr). Just putting this up so people can see what part of #17541 is already done. Will get back to fixing the test a little later.

### CODE REVIEW

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -59,6 +59,8 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
     # parse url for schema and host
     urlparts = match(urlmatcher, url)
     schema = urlparts.captures[1]
+    urlusername = urlparts.captures[4]
+    urlusername == nothing && (urlusername = "")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: `=== nothing`

File: base/libgit2/callbacks.jl [hunk 2]
```diff
@@ -111,15 +113,24 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
...
             else
-                if keydefpath === nothing || isempty(keydefpath)
-                    keydefpath = joinpath(homedir(),".ssh","id_rsa")
+                defaultkeydefpath = joinpath(homedir(),".ssh","id_rsa")
+                if isempty(keydefpath) && isfile(defaultkeydefpath)
+                    keydefpath = defaultkeydefpath
+                else
+                    keydefpath =
+                        prompt("Private key location for '$schema$username@$host'", default=keydefpath)
+                    @show keydefpath
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: to be removed before merging

File: base/libgit2/callbacks.jl [hunk 3]
```diff
@@ -142,11 +153,22 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
         end
         creds[:pubkey, credid] = publickey # save credentials
 
+        passphrase_required = true
+        if !isfile(privatekey)
+            warn("Private key `$privatekey` not found")
+        else
+            # In encrypted private keys, the second line is "Proc-Type: 4,ENCRYPTED"
+            open(privatekey) do f
+                passphrase_required = (readline(f); readline(f) == "Proc-Type: 4,ENCRYPTED\n")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: readchomp in case of crnl line endings?

File: test/TestHelpers.jl [hunk 4]
```diff
@@ -16,4 +16,31 @@ Base.Terminals.hascolor(t::FakeTerminal) = t.hascolor
...
+    rc = ccall(:grantpt, Cint, (Cint,), fdm)
+    rc != 0 && error("grantpt failed")
+    rc = ccall(:unlockpt, Cint, (Cint,), fdm)
+    rc != 0 && error("unlockpt")
+
+    fds = ccall(:open, Cint, (Ptr{UInt8}, Cint),
+        ccall(:ptsname, Ptr{UInt8}, (Cint,), fdm), O_RDWR|O_NOCTTY)
+
+    # slave
+    slave   = RawFD(fds)
```
Comment (tkelman) [inline test/TestHelpers.jl]: worker? child?

File: test/libgit2.jl [hunk 5]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
         @test creds[:pubkey, "localhost"] === nothing
         @test creds[:usesshagent, "localhost"] == "Y"
     #end
+
+    #@testset "SSH" begin
+        sshd_command = ""
+        ssh_repo = joinpath(dir, "Example.SSH")
+        !is_windows() && (sshd_command = strip(readstring(`which sshd`)))
```
Comment (tkelman) [inline test/libgit2.jl]: `which` is not always installed. how about ```success(`sshd`)``` inside a try catch?

File: test/libgit2.jl [hunk 6]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+        if !isempty(sshd_command)
+            mktempdir() do fakehomedir
+                oldhome = ENV["HOME"]
+                ENV["HOME"] = fakehomedir
```
Comment (tkelman) [inline test/libgit2.jl]: not needed due to withenv below?

File: test/libgit2.jl [hunk 7]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+                                reponame = "ssh://$(ENV["USER"])@localhost:$our_ssh_port$cache_repo"
+                                repo = LibGit2.clone(reponame, "$ssh_repo")
+                            finally
+                                finalize(repo)
+                            end
+                            """
+                            # We try to be helpful by desparately looking for
+                            # a way to prompt the password interactively. Pretend
+                            # to be a TTY to suppress those shenanigans. Further, we
+                            # need to detch since otherwise our terminal is still
```
Comment (tkelman) [inline test/libgit2.jl]: detach

File: test/libgit2.jl [hunk 8]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+                                end
+                                sleep(2)
+                                nb_available(master) != 0 &&
+                                    @show String(readavailable(master))
+                                @show p
+                                wait(p)
+                                nb_available(master) != 0 &&
+                                    @show String(readavailable(master))
+                            end
+                            run(`ls -la $ssh_repo`)
```
Comment (tkelman) [inline test/libgit2.jl]: not portable

File: test/libgit2.jl [hunk 9]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+
+                        # Ok, now encrypt the file and test with that (this also
+                        # makes sure that we don't accidentally fall back to the
+                        # unencrypted version)
+                        wait(spawn(`ssh-keygen -p -N "xxxxx" -f $fakehomedir/.ssh/id_rsa`))
+
+                        # Try with the encrypted file. Needs a password.
+                        try_clone(["Passphrase"=>"xxxxx\r\n"])
+
+                        # Move the file. It should now as for the location and
```
Comment (tkelman) [inline test/libgit2.jl]: ask

File: test/libgit2.jl [hunk 10]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+                        agent_sock = tempname()
+                        agentp = spawn(`ssh-agent -a $agent_sock -d`)
+                        while stat(agent_sock).mode == 0 # Wait until the agent is started
+                            sleep(1)
+                        end
+
+                        # fake pty is required for the same reason as in try_clone
+                        # above
+                        TestHelpers.with_fake_pty() do slave, master
+                            ENV["SSH_AUTH_SOCK"] = agent_sock
```
Comment (tkelman) [inline test/libgit2.jl]: withenv?

File: test/libgit2.jl [hunk 11]
```diff
@@ -599,6 +602,134 @@ mktempdir() do dir
...
+                                    slave, slave, slave)
+                            write(master, "xxxxx\n")
+                            wait(addp)
+                        end
+
+                        # Should now use the agent
+                        try_clone()
+                    finally
+                        p != nothing && kill(p)
+                        agentp != nothing && kill(agentp)
```
Comment (tkelman) [inline test/libgit2.jl]: `!== nothing`

File: base/libgit2/callbacks.jl [hunk 12]
```diff
@@ -59,6 +59,8 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
     # parse url for schema and host
     urlparts = match(urlmatcher, url)
     schema = urlparts.captures[1]
+    urlusername = urlparts.captures[4]
```
Comment (wildart) [inline base/libgit2/callbacks.jl]: usually, `username` from url is `git` for `git+ssh` protocol. You can use `username_ptr` parameter to get it.

Comment (Keno) [inline base/libgit2/callbacks.jl]: Sorry, not sure what you're suggesting. The `urlusername` only ever gets used by https. Are you saying that libgit2 is already extracting this, so we should just load it from `username_ptr`?

File: base/libgit2/callbacks.jl (hunk 3 above)
Comment (Keno) [inline base/libgit2/callbacks.jl]: `chomp(readline())` sounds reasonable

File: test/TestHelpers.jl (hunk 4 above)
Comment (Keno) [inline test/TestHelpers.jl]: This is the standard pty terminology (`http://man7.org/linux/man-pages/man7/pty.7.html`), I would not want to deviate.

**Comment by Keno** [comment]
I've rebased out the test and will open that as a separate pull request.

**Comment by tkelman** [comment]
move the WIP tests to a different PR?

File: base/libgit2/callbacks.jl [hunk 13]
```diff
@@ -114,15 +116,23 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
                 ENV["SSH_KEY_PATH"]
             else
                 keydefpath = creds[:prvkey, credid] # check if credentials were already used
-                if keydefpath !== nothing && !isusedcreds
+                keydefpath == nothing && (keydefpath = "")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: `=== nothing`

**Pull request merged**